skcosmo/pcovr/pcovr.py

```diff
@@ -0,0 +1,313 @@
+        space=None,
+        lr_args={"cv": 2, "fit_intercept": False},
+    ):
+
+        self.mixing = mixing
+        self.regularization = regularization
+        self.tol = tol
+        self.space = space
+        self.lr_args = lr_args
+        self.n_components = n_components
+        self.whiten = False
+        self._eig_solver = partial(
+            eig_solver, n_components=self.n_components, tol=self.tol, add_null=True
+        )
+
+    def fit(self, X, Y, Yhat=None, W=None):
+        """
+
+        Fit the model with X and Y. Depending on the dimensions of X,
+        calls either `_fit_feature_space` or `_fit_structure_space`
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training data, where n_samples is the number of samples and
+            n_features is the number of features.
+        Y : array-like, shape (n_samples, n_properties)
+            Training data, where n_samples is the number of samples and
+            n_properties is the number of properties
+        Yhat : array-like, shape (n_samples, n_properties), optional
+            Regressed training data, where n_samples is the number of samples and
+            n_properties is the number of properties. If not supplied, computed
+            by ridge regression.
+        W : array-like, shape (n_features, n_properties), optional
+            Weights of regressed training data. If not supplied, computed
+            by ridge regression.
+
+        """
+
+        X, Y = check_X_y(X, Y, y_numeric=True, multi_output=True)
+
+        if self.space is not None and self.space not in [
+            "feature",
+            "structure",
+            "auto",
+        ]:
+            raise ValueError("Only feature and structure space are supported.")
+
+        if self.n_components is None:
+            self.n_components = min(X.shape)
+
+        if Yhat is None or W is None:
+            Yhat, W = self._compute_Yhat(X, Y, Yhat=Yhat, W=W)
+
+        if self.space is None:
+            if X.shape[0] > X.shape[1]:
+                self.space = "feature"
+            else:
+                self.space = "structure"
+
+        if self.space == "feature":
+            self._fit_feature_space(X, Yhat, W)
+        else:
+            self._fit_structure_space(X, Yhat, W)
+
+        self.mean_ = np.mean(X, axis=0)
+        self.pxy_ = self.pxt_ @ self.pty_
+        if len(Y.shape) == 1:
+            self.pxy_ = self.pxy_.reshape(
+                X.shape[1],
+            )
+            self.pty_ = self.pty_.reshape(
+                self.n_components,
+            )
+
+        self.components_ = self.pxt_.T  # for sklearn compatibility
+
+    def _compute_Yhat(self, X, Y, Yhat=None, W=None):
+        """
+        Method for computing the approximation of Y to fit the PCovR
+        """
+
+        if Yhat is None:
+            if W is None:
+                lr = LR(**self.lr_args)  # some sort of args
+                lr.fit(X, Y)
```

Comment:
Looks like right now we are using a ridge regression model with built-in cross validation for the LR part -- I would argue it's better to just use a ridge regression model without automatic CV, especially if the number of CV folds is hard-coded, like it is here. Otherwise the user doesn't really have any control over the regularization used for the LR term (and the default CV is only 3 different regularization parameters, so the optimization will be really coarse in the current setup anyway).

I think better options are (1) let the user have complete control over the automatic CV in the LR part, or (2) make the user choose the LR hyperparameters manually, e.g., by doing cross-validation beforehand, and use `Ridge` instead of `RidgeCV`

Reply:
What if we allowed them to pass their own estimator, like Alex does in #9 ?

Reply:
Maybe, but then we run into the possibility that the user might make/use an estimator that isn't really compatible with PCovR, for example, using a kernel-based estimator in linear PCovR, or an SVM-based estimator. We could do this, but we would have to check that the estimator is a linear regression, which basically defeats the point of letting the user provide an estimator.

Reply:
So then I think we should leave RidgeCV as the default as it is most customizable, but set the default parameters to mimic LR. 

Right now the user can specify the lr arguments in the PCovR constructor, with the default being `dict(cv=2, fit_intercept=False)`. To mimic LR or Ridge one could just pass `lr_args = dict(alphas=[#])`, no?

Reply:
It should work, but my personal opinion is that the CV is still best left separate. I think it just confuses things a bit for the purpose of selecting one hyperparameter. If you want to keep the CV, it would be a good idea to then make accessible to the user all of the CV attributes as well -- otherwise it's a "black box" optimization and not particularly helpful.

Alternatively, why not make a `PCovRCV` class that can optimize the LR, the mixing, and the number of components? In my mind this would be a bit more consistent with the sklearn style, analogous to `Ridge` and `RidgeCV`.

Reply:
I like that idea. For now, I'll implement it with Ridge and constant regularization, and we can amend in a future PR.